File: go/82.remove-duplicates-from-sorted-list-ii.go
/*
 * @lc app=leetcode id=82 lang=golang
 *
 * [82] Remove Duplicates from Sorted List II
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteDuplicates(head *ListNode) *ListNode {
	return TwoPass(head)
}

func TwoPass(head *ListNode) *ListNode {
	count := make(map[int]int)
	current := head
	for current != nil {
		if _, exists := count[current.Val]; exists {
			count[current.Val]++
		} else {
			count[current.Val] = 1
		}
		current = current.Next
	}

	var prev *ListNode
	current = head
	for current != nil {
		if c := count[current.Val]; c > 1 {
			if prev == nil {
				head = current.Next
			} else {
				prev.Next = current.Next
			}
			current = current.Next
		} else {
			prev = current
			current = current.Next
		}
	}

	return head
}

File: csharp/82.remove-duplicates-from-sorted-list-ii.cs
/*
 * @lc app=leetcode id=82 lang=csharp
 *
 * [82] Remove Duplicates from Sorted List II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution 
{
    public ListNode DeleteDuplicates(ListNode head) 
    {
        if (head == null)
        {
            return null;
        }

        ListNode prev, cur, dummy;
        dummy = new ListNode(-101, head); // dummy head
        prev = dummy;
        cur = head;

        while (cur != null && cur.next != null)
        {
            // consecutive duplicates count
            int duplicateCount = 0;
            while (cur.next != null && cur.val == cur.next.val)
            {
                cur = cur.next;
                duplicateCount++;
            }

            // basically to find a edge. (edge triggerred)
            if (duplicateCount == 0)
            {
                prev = cur;
                cur = cur.next;
            }
            else 
            {
                prev.next = cur.next;
                cur = cur.next;
            }
        }

        return dummy.next;
    }
}
// @lc code=end

